### PR TITLE
Replace Thread.Sleep waits with ManualResetEventSlim

### PR DESCRIPTION
**Pull request by UselessToucan**
According to the PVS-Studio, these while loop based waitings are unreliable. See https://www.viva64.com/en/w/v3032.
Also, in my opinion they are ugly. :(

### CODE REVIEW

**Comment by peppy** [comment]
None of the cases in point fulfill the optimisation conditions of the linked document, FWIW.

File: osu.Framework.Tests/IO/BackgroundGameHeadlessGameHost.cs [hunk 1]
```diff
@@ -17,30 +17,38 @@ public class BackgroundGameHeadlessGameHost : HeadlessGameHost
...
                 }
                 catch
                 {
                     // may throw an unobserved exception if we don't handle here.
                 }
             });
 
-            while (testGame?.HasProcessed != true)
-                Thread.Sleep(10);
+            using (gameCreated)
```
Comment (peppy) [inline osu.Framework.Tests/IO/BackgroundGameHeadlessGameHost.cs]: please combine using with the construction.

File: osu.Framework.Tests/IO/BackgroundGameHeadlessGameHost.cs [hunk 2]
```diff
@@ -17,30 +17,38 @@ public class BackgroundGameHeadlessGameHost : HeadlessGameHost
...
+                gameCreated.Wait();
+
+            using (testGame.HasProcessed)
```
Comment (peppy) [inline osu.Framework.Tests/IO/BackgroundGameHeadlessGameHost.cs]: this is super dangerous. you are disposing something owned by a different class. just remove the using and put dispose logic in `TestGame`'s `Dispose`

**Comment by UselessToucan** [comment]
https://ci.appveyor.com/project/peppy/osu-framework/builds/25747109
image

**Comment by peppy** [comment]
please stop doing this. we are aware of. the failures via the automatic alerts.

**Pull request merged**